README.md
# Easy Toast
**easy toast** - is a vanilla javascript library for notifications.
<hr>
The purpose of this library is education.
I will be glad to your commits and code reviews in general!

The library weighs little, does not require css and is very easy to connect.

## Install via tag `<script>`:

Download `dist/toast.with.css.js` from [releases page](https://github.com/cr1gger/easy-toast/releases)

Connect to the head tag:
```html
<script src="toast.with.css.js"></script>
```

## Install via NPM:
Run:
```bash
npm install easy-v-toast
```
At the place of use:
```javascript
import Toast from "easy-v-toast/dist/toast.without.css";
import "easy-v-toast/dist/toast.min.css";

const toast = new EasyToast({
  // Configuration
})

```
## Usage

#### Default options and types notification
```javascript
toast.success("Hello it's easy success toast!")

toast.warning("Hello it's easy warning toast!")

toast.danger("Hello it's easy danger toast!")

toast.dark("Hello it's easy dark toast!")

toast.notify("Hello it's easy default toast!")

```

#### Custom options, if connected via tag `<script>`
```javascript
let config = {
    duration: 10,
    position: 'bottom-center',
    hasCloseButton: false
}

toast.configure(config)

toast.notify("Hello it's easy default toast, with custom options!") 

toast.notify("Hello it's easy success toast, with custom options!") 
```

#### Custom options, if connected via NPM
```javascript
let config = {
    duration: 5,
    events: {
        click: (e) => {
            console.log('Your event click by notification!')
        }
    }
}

const toast = new EasyToast(config)

toast.dark('Dark notification, with custom options and events')
```

## Configuration
```javascript
const defaultConfig = {
    duration: 3,
    position: 'bottom-right',
    hasCloseButton: true,
    showTimeline: true,
    autoclose: true,
    stopOnHover: true,
    events: {
        'click': null,
        'closeClick': null,
    }
}
```

#### There are two ways to set the configuration:

First via constructor:
```javascript
  const toast = new EasyToast(config)
```

The second way, through the configure method:
```javascript
  toast.configure(config)
```

#### Available properties:
- `duration` - Duration in seconds
- `position` - top-right
  | top-middle
  | top-left
  | bottom-left
  | bottom-middle
  | bottom-right
- `hasCloseButton` - Visible close button | boolean
- `showTimeline` - Visible progress bar | boolean
- `autoclose` - Autoclose after the end of the timer | boolean
- `stopOnHover` - Pause the timer when you hover over a notification | boolean
- `events` - Object contains two available event: click and closeClick

### Version: 1.0.0


CHANGELOG.md
# Change log

### 1.0.0
- Initial version


demo/js/toast.with.css.js
(()=>{"use strict";var t={380:(t,e,o)=>{o.d(e,{Z:()=>i});var n=o(81),r=o.n(n),a=o(645),s=o.n(a)()(r());s.push([t.id,'svg.easy-toast{display:block;vertical-align:middle}*.easy-toast,.easy-toast::before,.easy-toast::after{z-index:999999;box-sizing:border-box;border-width:0;border-style:solid;border-color:#e5e7eb;line-height:1.5;font-family:ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"}p.easy-toast{margin:0}.easy-toast-block{max-width:20rem;font-size:0.875rem;line-height:1.25rem;border-radius:0.375rem;box-shadow:0 10px 15px -3px rgba(0,0,0,0.3),0 4px 6px -4px rgba(0,0,0,0.3);color:white;margin-top:0.5rem;word-break:break-all}.easy-toast-stack{display:flex;flex-direction:column;margin-top:0.875rem;margin-bottom:0.875rem}.easy-toast-button{display:inline-flex;flex-shrink:0;align-items:center;height:1rem;width:1rem;border-radius:0.375rem;color:rgba(255,255,255,0.5) !important;font-size:0.875rem;line-height:1.25rem;transition-property:all;transition-timing-function:cubic-bezier(0.4, 0, 0.2, 1);transition-duration:150ms;cursor:pointer;text-shadow:none;text-align:center;-webkit-appearance:button;background-color:transparent;background-image:none;margin:0;padding:0;writing-mode:horizontal-tb !important}.easy-toast-button:hover{color:white !important}.easy-toast-svg-close{width:0.875rem;height:0.875rem}.easy-toast-position-top-left{top:1rem;left:1rem}.easy-toast-position-top-middle{top:1rem;left:50%;transform:translate(-50%, 0)}.easy-toast-position-top-right{top:1rem;right:1rem}.easy-toast-position-bottom-left{bottom:1rem;left:1rem}.easy-toast-position-bottom-middle{bottom:1rem;left:50%;transform:translate(-50%, 0)}.easy-toast-position-bottom-right{bottom:1rem;right:1rem}.easy-toast-position-center{top:50%;left:50%;transform:translate(-50%, -50%)}.easy-toast-timeline{border-radius:0.375rem;width:100%;height:0.25rem;opacity:0.5}.et-bg-gray-500{background-color:#6b7280}.et-bg-green-500{background-color:#22c55e}.et-bg-red-500{background-color:#ef4444}.et-bg-yellow-500{background-color:#eab308}.et-bg-green-900{background-color:#14532d}.et-bg-red-900{background-color:#7f1d1d}.et-bg-yellow-900{background-color:#713f12}.et-bg-gray-900{background-color:#111827}.et-flex{display:flex}.et-inline-flex{display:inline-flex}.et-flex-shrink-0{flex-shrink:0}.et-p-4{padding:1rem}.et-px-3{padding-left:0.75rem;padding-right:0.75rem}.et-ml-auto{margin-left:auto}.et-fixed{position:fixed}.et-sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0, 0, 0, 0);white-space:nowrap;border-width:0}\n',""]);const i=s},645:t=>{t.exports=function(t){var e=[];return e.toString=function(){return this.map((function(e){var o="",n=void 0!==e[5];return e[4]&&(o+="@supports (".concat(e[4],") {")),e[2]&&(o+="@media ".concat(e[2]," {")),n&&(o+="@layer".concat(e[5].length>0?" ".concat(e[5]):""," {")),o+=t(e),n&&(o+="}"),e[2]&&(o+="}"),e[4]&&(o+="}"),o})).join("")},e.i=function(t,o,n,r,a){"string"==typeof t&&(t=[[null,t,void 0]]);var s={};if(n)for(var i=0;i<this.length;i++){var c=this[i][0];null!=c&&(s[c]=!0)}for(var l=0;l<t.length;l++){var u=[].concat(t[l]);n&&s[u[0]]||(void 0!==a&&(void 0===u[5]||(u[1]="@layer".concat(u[5].length>0?" ".concat(u[5]):""," {").concat(u[1],"}")),u[5]=a),o&&(u[2]?(u[1]="@media ".concat(u[2]," {").concat(u[1],"}"),u[2]=o):u[2]=o),r&&(u[4]?(u[1]="@supports (".concat(u[4],") {").concat(u[1],"}"),u[4]=r):u[4]="".concat(r)),e.push(u))}},e}},81:t=>{t.exports=function(t){return t[1]}},379:t=>{var e=[];function o(t){for(var o=-1,n=0;n<e.length;n++)if(e[n].identifier===t){o=n;break}return o}function n(t,n){for(var a={},s=[],i=0;i<t.length;i++){var c=t[i],l=n.base?c[0]+n.base:c[0],u=a[l]||0,p="".concat(l," ").concat(u);a[l]=u+1;var d=o(p),m={css:c[1],media:c[2],sourceMap:c[3],supports:c[4],layer:c[5]};if(-1!==d)e[d].references++,e[d].updater(m);else{var f=r(m,n);n.byIndex=i,e.splice(i,0,{identifier:p,updater:f,references:1})}s.push(p)}return s}function r(t,e){var o=e.domAPI(e);return o.update(t),function(e){if(e){if(e.css===t.css&&e.media===t.media&&e.sourceMap===t.sourceMap&&e.supports===t.supports&&e.layer===t.layer)return;o.update(t=e)}else o.remove()}}t.exports=function(t,r){var a=n(t=t||[],r=r||{});return function(t){t=t||[];for(var s=0;s<a.length;s++){var i=o(a[s]);e[i].references--}for(var c=n(t,r),l=0;l<a.length;l++){var u=o(a[l]);0===e[u].references&&(e[u].updater(),e.splice(u,1))}a=c}}},569:t=>{var e={};t.exports=function(t,o){var n=function(t){if(void 0===e[t]){var o=document.querySelector(t);if(window.HTMLIFrameElement&&o instanceof window.HTMLIFrameElement)try{o=o.contentDocument.head}catch(t){o=null}e[t]=o}return e[t]}(t);if(!n)throw new Error("Couldn't find a style target. This probably means that the value for the 'insert' parameter is invalid.");n.appendChild(o)}},216:t=>{t.exports=function(t){var e=document.createElement("style");return t.setAttributes(e,t.attributes),t.insert(e,t.options),e}},565:(t,e,o)=>{t.exports=function(t){var e=o.nc;e&&t.setAttribute("nonce",e)}},795:t=>{t.exports=function(t){var e=t.insertStyleElement(t);return{update:function(o){!function(t,e,o){var n="";o.supports&&(n+="@supports (".concat(o.supports,") {")),o.media&&(n+="@media ".concat(o.media," {"));var r=void 0!==o.layer;r&&(n+="@layer".concat(o.layer.length>0?" ".concat(o.layer):""," {")),n+=o.css,r&&(n+="}"),o.media&&(n+="}"),o.supports&&(n+="}");var a=o.sourceMap;a&&"undefined"!=typeof btoa&&(n+="\n/*# sourceMappingURL=data:application/json;base64,".concat(btoa(unescape(encodeURIComponent(JSON.stringify(a))))," */")),e.styleTagTransform(n,t,e.options)}(e,t,o)},remove:function(){!function(t){if(null===t.parentNode)return!1;t.parentNode.removeChild(t)}(e)}}}},589:t=>{t.exports=function(t,e){if(e.styleSheet)e.styleSheet.cssText=t;else{for(;e.firstChild;)e.removeChild(e.firstChild);e.appendChild(document.createTextNode(t))}}}},e={};function o(n){var r=e[n];if(void 0!==r)return r.exports;var a=e[n]={id:n,exports:{}};return t[n](a,a.exports,o),a.exports}o.n=t=>{var e=t&&t.__esModule?()=>t.default:()=>t;return o.d(e,{a:e}),e},o.d=(t,e)=>{for(var n in e)o.o(e,n)&&!o.o(t,n)&&Object.defineProperty(t,n,{enumerable:!0,get:e[n]})},o.o=(t,e)=>Object.prototype.hasOwnProperty.call(t,e),o.nc=void 0,(()=>{const t="easy-toast-stack-br",e="easy-toast-timeline",n="top-right",r="top-middle",a="top-left",s="bottom-left",i="bottom-middle",c="bottom-right",l="center",u="default",p="success",d="danger",m="warning",f="dark",g={duration:3,type:"default",position:"bottom-right",hasCloseButton:!0,showTimeline:!0,autoclose:!0,stopOnHover:!0,events:{click:null,closeClick:null}};function y(o,y={}){let v={...g,...y},w=function(e){let o=function(t){switch(t){case a:return"easy-toast-position-top-left";case r:return"easy-toast-position-top-middle";case n:return"easy-toast-position-top-right";case s:return"easy-toast-position-bottom-left";case i:return"easy-toast-position-bottom-middle";case c:return"easy-toast-position-bottom-right";case l:return"easy-toast-position-center";default:return"easy-toast-position-bottom-right"}}(e),u=function(e){switch(e){case a:return"easy-toast-stack-tl";case r:return"easy-toast-stack-tm";case n:return"easy-toast-stack-tr";case s:return"easy-toast-stack-bl";case i:return"easy-toast-stack-bm";case c:return t;case l:return"easy-toast-stack-center";default:return t}}(e),p=document.querySelector("."+u);if(!p){let t=document.createElement("div");t.className="easy-toast et-fixed "+o,p=document.createElement("div"),p.className="easy-toast easy-toast-stack "+u,t.append(p),document.body.append(t)}return p}(v.position),{notify:x,notifyClose:k,notifyTimeline:C}=function(t,o){let n=function(t){switch(t){case u:return"et-bg-gray-500";case p:return"et-bg-green-500";case d:return"et-bg-red-500";case m:return"et-bg-yellow-500";case f:return"et-bg-gray-900"}}(o.type),r=function(t){switch(t){case u:return"et-bg-gray-900";case p:return"et-bg-green-900";case d:return"et-bg-red-900";case m:return"et-bg-yellow-900";case f:return"et-bg-gray-500"}}(o.type),a=document.createElement("div");a.className="easy-toast easy-toast-block "+n;let s=document.createElement("div");s.className="easy-toast et-p-4 et-flex";let i=document.createElement("p");i.className="easy-toast et-px-3",i.innerHTML=t.toString().replaceAll("script","");let c=document.createElement("div");c.className="easy-toast et-ml-auto";let l=document.createElement("button");l.className="easy-toast easy-toast-button",l.innerHTML='\n        <span class="easy-toast et-sr-only">Close</span>\n        <svg class="easy-toast easy-toast-svg-close" width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">\n            <path d="M0.92524 0.687069C1.126 0.486219 1.39823 0.373377 1.68209 0.373377C1.96597 0.373377 2.2382 0.486219 2.43894 0.687069L8.10514 6.35813L13.7714 0.687069C13.8701 0.584748 13.9882 0.503105 14.1188 0.446962C14.2494 0.39082 14.3899 0.361248 14.5321 0.360026C14.6742 0.358783 14.8151 0.38589 14.9468 0.439762C15.0782 0.493633 15.1977 0.573197 15.2983 0.673783C15.3987 0.774389 15.4784 0.894026 15.5321 1.02568C15.5859 1.15736 15.6131 1.29845 15.6118 1.44071C15.6105 1.58297 15.5809 1.72357 15.5248 1.85428C15.4688 1.98499 15.3872 2.10324 15.2851 2.20206L9.61883 7.87312L15.2851 13.5441C15.4801 13.7462 15.588 14.0168 15.5854 14.2977C15.5831 14.5787 15.4705 14.8474 15.272 15.046C15.0735 15.2449 14.805 15.3574 14.5244 15.3599C14.2437 15.3623 13.9733 15.2543 13.7714 15.0591L8.10514 9.38812L2.43894 15.0591C2.23704 15.2543 1.96663 15.3623 1.68594 15.3599C1.40526 15.3574 1.13677 15.2449 0.938279 15.046C0.739807 14.8474 0.627232 14.5787 0.624791 14.2977C0.62235 14.0168 0.730236 13.7462 0.92524 13.5441L6.59144 7.87312L0.92524 2.20206C0.724562 2.00115 0.611816 1.72867 0.611816 1.44457C0.611816 1.16047 0.724562 0.887983 0.92524 0.687069Z" fill="currentColor"/>\n        </svg>\n        ';let g=document.createElement("div");return g.className=`easy-toast ${r} ${e}`,s.append(i),o.hasCloseButton&&c.append(l),s.append(c),a.append(s),o.showTimeline&&o.autoclose&&a.append(g),{notify:a,notifyClose:l,notifyTimeline:g}}(o,v);v.stopOnHover&&(x.onmouseenter=function(){x.setAttribute("data-pause","pause")},x.onmouseleave=function(){x.removeAttribute("data-pause");let t=x.getAttribute("data-start"),e=x.getAttribute("data-last");b(x,v,e-t,C.style.width.replace("%",""))}),w.prepend(x);let A=b(x,v);v.hasCloseButton&&("function"==typeof v.events.closeClick?k.onclick=function(t){t.stopPropagation(),v.events.closeClick(t)}:k.onclick=t=>{t.stopPropagation(),h(x,A)}),"function"==typeof v.events.click&&(x.onclick=function(t){t.stopPropagation(),v.events.click(t)})}function b(t,o,n=null,r=null){t.setAttribute("data-start",Date.now());let a=n?1e3*o.duration-n:1e3*o.duration,s=r||100,i=v({timing:function(t){return s-100*t},draw:n=>{o.showTimeline&&o.autoclose&&(t.setAttribute("data-last",Date.now()),t.querySelector(`.${e}`).style.width=n+"%"),n<1&&o.autoclose&&h(t,i)},duration:a,notify:t});return i}function h(t,e=null){t.remove(),e&&cancelAnimationFrame(e)}const v=({timing:t,draw:e,duration:o,notify:n})=>{let r=performance.now();return requestAnimationFrame((function a(s){if(n.getAttribute("data-pause"))return;let i=(s-r)/o;i>1&&(i=1);let c=t(i);return e(c),i<1?requestAnimationFrame(a):void 0}))};function w(t={}){this.configuration=t}w.prototype.success=function(t){y(t,{...this.configuration,type:"success"})},w.prototype.warning=function(t){y(t,{...this.configuration,type:"warning"})},w.prototype.danger=function(t){y(t,{...this.configuration,type:"danger"})},w.prototype.dark=function(t){y(t,{...this.configuration,type:"dark"})},w.prototype.notify=function(t){y(t,{...this.configuration,type:"default"})},w.prototype.configure=function(t){this.configuration=t};const x=w;var k=o(379),C=o.n(k),A=o(795),E=o.n(A),T=o(569),S=o.n(T),N=o(565),L=o.n(N),M=o(216),I=o.n(M),H=o(589),j=o.n(H),P=o(380),z={};z.styleTagTransform=j(),z.setAttributes=L(),z.insert=S().bind(null,"head"),z.domAPI=E(),z.insertStyleElement=I(),C()(P.Z,z),P.Z&&P.Z.locals&&P.Z.locals;const F=new x({position:"bottom-right",duration:10});globalThis.toast=F,F.success("Hello from index")})()})();

src/js/toast.js
// Stack position classes
const STACK_CLASS_TL = 'easy-toast-stack-tl'
const STACK_CLASS_TR = 'easy-toast-stack-tr'
const STACK_CLASS_TM = 'easy-toast-stack-tm'
const STACK_CLASS_BR = 'easy-toast-stack-br'
const STACK_CLASS_BL = 'easy-toast-stack-bl'
const STACK_CLASS_BM = 'easy-toast-stack-bm'
const STACK_CLASS_CENTER = 'easy-toast-stack-center'
const TIMELINE_CLASS = 'easy-toast-timeline'

// Positions
const POSITION_TR = 'top-right'
const POSITION_TM = 'top-middle'
const POSITION_TL = 'top-left'
const POSITION_BL = 'bottom-left'
const POSITION_BM = 'bottom-middle'
const POSITION_BR = 'bottom-right'
const POSITION_CENTER = 'center'

// Types
const TYPE_DEFAULT = 'default'
const TYPE_SUCCESS = 'success'
const TYPE_DANGER = 'danger'
const TYPE_WARNING = 'warning'
const TYPE_DARK = 'dark'

const DEFAULT_CONFIG = {
    duration: 3,
    type: 'default',
    position: 'bottom-right',
    hasCloseButton: true,
    showTimeline: true,
    autoclose: true,
    stopOnHover: true,
    events: {
        'click': null,
        'closeClick': null,
    }
}

function runNotify(message, cfg = {}) {
    let config = {...DEFAULT_CONFIG, ...cfg }

    let stack = getStack(config.position)
    let { notify, notifyClose, notifyTimeline } = createNotify(message, config)

    if (config.stopOnHover) {
        notify.onmouseenter = function() {
            notify.setAttribute('data-pause', 'pause');
        }
        notify.onmouseleave = function() {
            notify.removeAttribute('data-pause');
            let startTime = notify.getAttribute('data-start')
            let lastTime = notify.getAttribute('data-last')

            startAnimate(notify, config, (lastTime - startTime), notifyTimeline.style.width.replace('%', ''))
        }
    }

    stack.prepend(notify)

    let animateId = startAnimate(notify, config)

    // Register event close click
    if (config.hasCloseButton) {
        if (typeof config.events.closeClick === 'function')
        {
            notifyClose.onclick = function(e) {
                e.stopPropagation();
                config.events.closeClick(e);
            }
        } else {
            notifyClose.onclick = (e) => {
                e.stopPropagation();
                destroy(notify, animateId)
            }
        }
    }

    // Register event notify click
    if (typeof config.events.click === 'function')
    {
        notify.onclick = function(e) {
            e.stopPropagation();
            config.events.click(e);
        }
    }
}

function startAnimate(notify, config, startDuration = null, startPercent = null) {
    notify.setAttribute('data-start', Date.now())

    let duration = startDuration ?
        config.duration * 1000 - startDuration : config.duration * 1000

    let percent = nullOperation(startPercent, 100)


    let animateId = animate({
        timing: function(timeFraction) {
            return percent - (timeFraction * 100)
        },
        draw: (progress) => {
            if (config.showTimeline && config.autoclose) {
                notify.setAttribute('data-last', Date.now())
                notify.querySelector(`.${TIMELINE_CLASS}`).style.width = progress + '%';
            }
            if (progress < 1) {
                if (config.autoclose) {
                    destroy(notify, animateId)
                }
            }
        },
        duration,
        notify
    })

    return animateId;
}

function nullOperation(value, defaultValue) {
    if (value) {
        return value;
    }
    if (defaultValue)
    {
        return defaultValue
    }

    return null;
}

function createNotify(text, config) {
    let typeClass = getTypeClass(config.type)
    let timeLineBg = getTimelineClass(config.type)

    let notifyBlock = document.createElement('div');
    notifyBlock.className = 'easy-toast easy-toast-block ' + typeClass

    let notifyFlex = document.createElement('div');
    notifyFlex.className = 'easy-toast et-p-4 et-flex'

    let notifyTitle = document.createElement('p')
    notifyTitle.className = 'easy-toast et-px-3'
    notifyTitle.innerHTML = text.toString().replaceAll('script', '')

    let notifyContainer = document.createElement('div')
    notifyContainer.className = 'easy-toast et-ml-auto'

    let notifyClose = document.createElement('button')
    notifyClose.className = 'easy-toast easy-toast-button'

    notifyClose.innerHTML = `
        <span class="easy-toast et-sr-only">Close</span>
        <svg class="easy-toast easy-toast-svg-close" width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
            <path d="M0.92524 0.687069C1.126 0.486219 1.39823 0.373377 1.68209 0.373377C1.96597 0.373377 2.2382 0.486219 2.43894 0.687069L8.10514 6.35813L13.7714 0.687069C13.8701 0.584748 13.9882 0.503105 14.1188 0.446962C14.2494 0.39082 14.3899 0.361248 14.5321 0.360026C14.6742 0.358783 14.8151 0.38589 14.9468 0.439762C15.0782 0.493633 15.1977 0.573197 15.2983 0.673783C15.3987 0.774389 15.4784 0.894026 15.5321 1.02568C15.5859 1.15736 15.6131 1.29845 15.6118 1.44071C15.6105 1.58297 15.5809 1.72357 15.5248 1.85428C15.4688 1.98499 15.3872 2.10324 15.2851 2.20206L9.61883 7.87312L15.2851 13.5441C15.4801 13.7462 15.588 14.0168 15.5854 14.2977C15.5831 14.5787 15.4705 14.8474 15.272 15.046C15.0735 15.2449 14.805 15.3574 14.5244 15.3599C14.2437 15.3623 13.9733 15.2543 13.7714 15.0591L8.10514 9.38812L2.43894 15.0591C2.23704 15.2543 1.96663 15.3623 1.68594 15.3599C1.40526 15.3574 1.13677 15.2449 0.938279 15.046C0.739807 14.8474 0.627232 14.5787 0.624791 14.2977C0.62235 14.0168 0.730236 13.7462 0.92524 13.5441L6.59144 7.87312L0.92524 2.20206C0.724562 2.00115 0.611816 1.72867 0.611816 1.44457C0.611816 1.16047 0.724562 0.887983 0.92524 0.687069Z" fill="currentColor"/>
        </svg>
        `


    let notifyTimeline = document.createElement('div')
    notifyTimeline.className = `easy-toast ${timeLineBg} ${TIMELINE_CLASS}`

    notifyFlex.append(notifyTitle)

    if (config.hasCloseButton) {
        notifyContainer.append(notifyClose)
    }

    notifyFlex.append(notifyContainer)
    notifyBlock.append(notifyFlex)

    if (config.showTimeline && config.autoclose) {
        notifyBlock.append(notifyTimeline)
    }

    return {
        notify:notifyBlock,
        notifyClose,
        notifyTimeline
    };

}

function getStack(position) {
    let positionClass = getPositionClasses(position)
    let stackClass = getStackClass(position)

    let stack = document.querySelector('.' + stackClass)
    if (!stack)
    {
        let stackContainer = document.createElement('div')
        stackContainer.className = 'easy-toast et-fixed ' + positionClass

        stack = document.createElement('div')
        stack.className = 'easy-toast easy-toast-stack ' + stackClass;

        stackContainer.append(stack)

        document.body.append(stackContainer)
    }
    return stack;
}

function destroy(block, animateId = null) {
    block.remove()
    if (animateId) {
        cancelAnimationFrame(animateId)
    }
}

function getPositionClasses(position) {
    switch (position) {
        case POSITION_TL: return 'easy-toast-position-top-left';
        case POSITION_TM: return 'easy-toast-position-top-middle';
        case POSITION_TR: return 'easy-toast-position-top-right';

        case POSITION_BL: return 'easy-toast-position-bottom-left';
        case POSITION_BM: return 'easy-toast-position-bottom-middle';
        case POSITION_BR: return 'easy-toast-position-bottom-right';
        case POSITION_CENTER: return 'easy-toast-position-center';
        default: return 'easy-toast-position-bottom-right'

    }
}

function getStackClass(position) {
    switch (position) {
        case POSITION_TL: return STACK_CLASS_TL;
        case POSITION_TM: return STACK_CLASS_TM;
        case POSITION_TR: return STACK_CLASS_TR;
        case POSITION_BL: return STACK_CLASS_BL;
        case POSITION_BM: return STACK_CLASS_BM;
        case POSITION_BR: return STACK_CLASS_BR;
        case POSITION_CENTER: return STACK_CLASS_CENTER;
        default: return STACK_CLASS_BR;
    }
}

function getTypeClass(type) {
    switch (type) {
        case TYPE_DEFAULT: return 'et-bg-gray-500';
        case TYPE_SUCCESS: return 'et-bg-green-500';
        case TYPE_DANGER: return 'et-bg-red-500';
        case TYPE_WARNING: return 'et-bg-yellow-500';
        case TYPE_DARK: return 'et-bg-gray-900';
    }
}

function getTimelineClass(type) {
    switch (type) {
        case TYPE_DEFAULT: return 'et-bg-gray-900';
        case TYPE_SUCCESS: return 'et-bg-green-900';
        case TYPE_DANGER: return 'et-bg-red-900';
        case TYPE_WARNING: return 'et-bg-yellow-900';
        case TYPE_DARK: return 'et-bg-gray-500';
    }
}

const animate = ({timing, draw, duration, notify}) => {

    let start = performance.now();

    return requestAnimationFrame(function animate(time) {

        if (notify.getAttribute('data-pause')) return;

        // timeFraction изменяется от 0 до 1
        let timeFraction = (time - start) / duration;
        if (timeFraction > 1) timeFraction = 1;
        // вычисление текущего состояния анимации
        let progress = timing(timeFraction);

        draw(progress); // отрисовать её

        if (timeFraction < 1) {
            return requestAnimationFrame(animate);
        }

    });
}


function Toast (configuration = {}) {

    this.configuration = configuration
}

Toast.prototype.success = function(message) {
    let config = {...this.configuration, type: 'success' }
    runNotify(message, config);
}

Toast.prototype.warning = function(message) {
    let config = {...this.configuration, type: 'warning' }
    runNotify(message, config);
}

Toast.prototype.danger = function(message) {
    let config = {...this.configuration, type: 'danger' }
    runNotify(message, config);
}

Toast.prototype.dark = function(message) {
    let config = {...this.configuration, type: 'dark' }
    runNotify(message, config);
}

Toast.prototype.configure = function(config) {
    this.configuration = config
}

Toast.prototype.notify = function(message) {
    let config = {...this.configuration, type: 'default' }
    runNotify(message, config)
}

export default Toast
